openat2_linux.go

```diff
@@ -0,0 +1,105 @@
+//go:build linux
+
+// Copyright (C) 2024 SUSE LLC. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package securejoin
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	hasOpenat2Bool bool
+	hasOpenat2Once sync.Once
+)
+
+func hasOpenat2() bool {
+	hasOpenat2Once.Do(func() {
+		fd, err := unix.Openat2(unix.AT_FDCWD, ".", &unix.OpenHow{
+			Flags:   unix.O_PATH | unix.O_CLOEXEC,
+			Resolve: unix.RESOLVE_NO_SYMLINKS | unix.RESOLVE_IN_ROOT,
+		})
+		if err == nil {
+			hasOpenat2Bool = true
+			_ = unix.Close(fd)
+		}
+	})
+	return hasOpenat2Bool
+}
+
+func openat2File(dir *os.File, path string, how *unix.OpenHow) (*os.File, error) {
+	fullPath := dir.Name() + "/" + path
+	// Make sure we always set O_CLOEXEC.
+	how.Flags |= unix.O_CLOEXEC
+	var tries int
+	for tries < 10 {
+		fd, err := unix.Openat2(int(dir.Fd()), path, how)
+		if err != nil {
+			if err == unix.EAGAIN {
```

Comment:
`RESOLVE_IN_ROOT` can return `-EAGAIN` if there was a rename or mount on the system while resolving `..` (this is because the change in the tree could make `..` unsafe, and doing the retry in the kernel could result in a DoS, so the kernel API lets userspace decide if they want to do the retry).

Reply:
The mount needs to happen beneath the root or if it happens somewhere beneath "/" we can see EAGAIN?

I'm ok if this is not properly documented and we want to be sure, though.

Reply:
Anywhere on the system. All mounts and renames on Linux are tied to a global seqlock. See https://github.com/torvalds/linux/blob/v6.9/fs/namei.c#L1970. The exact cases where `-EAGAIN` is returned isn't described in the man page because it's an implementation detail of how scoped lookups are implemented.

Reply:
Cool. I'll add a comment, the manpage is not very clear about this case.